### PR TITLE
gap: update to 4.12.1

### PR DESCRIPTION
**Pull request by fingolfin**
I am not a Termux user myself, so I am not completely sure I got this right.
(Rather, I am one of the developers of GAP.) Apologies if I did something
wrong here.

### CODE REVIEW

File: packages/gap/build.sh [hunk 1]
```diff
@@ -2,10 +2,10 @@ TERMUX_PKG_HOMEPAGE=https://www.gap-system.org/
 TERMUX_PKG_DESCRIPTION="GAP is a system for computational discrete algebra, with particular emphasis on Computational Group Theory"
 TERMUX_PKG_LICENSE="GPL-2.0"
 TERMUX_PKG_MAINTAINER="@termux"
-TERMUX_PKG_VERSION=4.11.1
-TERMUX_PKG_REVISION=1
+TERMUX_PKG_VERSION=4.12.1
+TERMUX_PKG_REVISION=0
```
Comment (thunder-coding) [inline packages/gap/build.sh:6]: Simply remove this line when updating to newer version. No need to have it set to zero.

**Review by thunder-coding — CHANGES_REQUESTED** [review]

File: packages/gap/build.sh (hunk 1 above)
Comment (fingolfin) [reply packages/gap/build.sh]: Fixed

**Comment by fingolfin** [comment]
Unfortunately, I noticed that this packages seems mostly useless in its current form, as it seems to skip completely over installing GAP packages (included in the `pkg` subdirectory of the tarball). I've also noticed you have a rule for keeping packages below 100 MB; I was wondering how this recipe managed that; well, if one skips `pkg`, that's easy. But at the same time, there are currently four "required" packages; GAP won't even start without those.

I don't know a way to try out Termux packages right now, but I wonder if the current `gap` packages even allows starting `gap`?

**Comment by fingolfin** [comment]
Found https://github.com/termux/termux-docker and tried it, and yeah, it is exactly as I expected:
```
Preparing to unpack .../gap_4.11.1-1_x86%5f64.deb ...
Unpacking gap (4.11.1-1) ...
Setting up gap (4.11.1-1) ...
~ $ gap
 ┌───────┐   GAP 4.11.1 of 2021-03-02
 │  GAP  │   https://www.gap-system.org
 └───────┘   Architecture: x86_64-pc-linux-android-default64-kv7
 Configuration:  gmp 6.2.1, GASMAN, readline
 Loading the library and packages ...
#I  gapdoc package is not available. Check that the name is correct
#I  and it is present in one of the GAP root directories (see '??RootPaths')
#I  GAP: needed package gapdoc cannot be loaded
Error, failed to load needed package `gapdoc' (version >= 1.2) at /data/data/com.termux/files/usr/share/gap/lib/package.gi:1919 called from
func(  ); at /data/data/com.termux/files/usr/share/gap/lib/system.g:216 called from
<function "CallAndInstallPostRestore">( <arguments> )
 called from read-eval loop at /data/data/com.termux/files/usr/share/gap/lib/init.g:690
you can 'quit;' to quit to outer loop, or
you can 'return;' to continue
brk>
```
(Although it is not completely useless as one can start it via `gap --bare`, but in that mode all kinds of things won't work right, it is mostly for bootstrapping)

I could address this here, but installing all packages would probably end up using > 1 GB disk space. Even with the minimal set of packages and some tricks I don't think one can stay under the 100 MB limit (but < 200 MB may be possible)

**Comment by fingolfin** [comment]
The build fails because it tries to execute `ffgen` which fails (due to cross compilation). But this is weird, it *shouldn't* execute `ffgen` (precisely to enable cross compilation), that tool should only be run in git snapshots of GAP (source: I wrote that code, and regularly cross compile it, too). But of course this could be a regression in 4.12.1... alas I just verified it locally (downloaded the GAP tarball referenced in this PR, run `./configure && make` in it -- no `ffgen` binary was even created, let alone run. Weird...

**Comment by fingolfin** [comment]
Regarding packages: note that gapdoc+smallgrp+transgrp+primgrp in unmodified form already take > 300 MB storage, *but* one can use that GAP supports transparent *.gz compression and run something like  `gzip -9 -n -f transgrp/data/*.* primgrp/data/*.* smallgrp/id*/*.* smallgrp/small*/*.*` to get them "small enough". (The `-f` to avoid issues with the files with extension `.z` -- those are *not* compressed files, they are all plain text. And the `-n` to improve reproducibility, by avoiding pointless timestamps in the `.gz`)

**Comment by fingolfin** [comment]
Ahh, wait, I just noticed the line `configure: out-of-tree build` in the build log... and now I understand the problem: the code for *not* regenerating `ffdata.c` when it is already present in `src` does not take out-of-tree builds into account. So this is indeed a bug on our (GAP) side. I'll work on a fix for the next release, which then can also be added as a patch here.

**Comment by fingolfin** [comment]
I've now added a workaround for the `ffgen` issue, and some (currently commented out) code for installing a minimal set of packages. Someone needs to let the CI tests run again, though :-)

**Comment by Grimler91** [comment]
Added some commits to avoid `install -s`, and to create a gap-packages subpackage, gap-packages for aarch64 is 91 MB, and gap 22 MB, so acceptable sizes

**Comment by fingolfin** [comment]
Thanks, that looks fine to me! I'd be happy if this PR was merged as-is, overall it should be a net-positive for anyone trying to use GAP this way.


Down the road, perhaps one could also wrap the remaining GAP packages (in the `pkg` dir),  each in a separate (sub-?)package -- it wouldn't be too hard to write a script which takes the list of all packages and their metadata (available at <https://github.com/gap-system/gap/releases/download/v4.12.1/package-infos.json.gz>) and spits out code for each (sub-)package. In other distros, typically a GAP package `foo` will be wrapped as `gap-pkg-foo` or so (to avoid clashes with other software). But perhaps "subpackages" are not the right way to do it, as each of those has a version of its own, independent of that from GAP... Anyway, I am not a Termux user, so I won't delve into this, I just wanted to mention the possibility :-).

**Comment by Grimler91** [comment]
Thanks!

**Pull request merged**